src/mgr/DaemonServer.cc

```diff
@@ -924,6 +924,70 @@ bool DaemonServer::handle_command(MCommand *m)
       });
     cmdctx->reply(r, "");
     return true;
+  } else if (prefix == "osd safe-to-destroy") {
+    int64_t osd;
+    if (!cmd_getval(g_ceph_context, cmdctx->cmdmap, "id", osd)) {
+      ss << "unable to parse osd id value '"
+         << cmd_vartype_stringify(cmdctx->cmdmap["id"]) << "'";
+      cmdctx->reply(-EINVAL, ss);
+      return true;
+    }
+    bool exists;
+    cluster_state.with_osdmap([&](const OSDMap& osdmap) {
+	exists = osdmap.exists(osd);
+      });
+    if (!exists) {
+      ss << "osd." << osd << " does not exist";
+      cmdctx->reply(0, ss);
+      return true;
+    }
+    int r = 0;
+    cluster_state.with_pgmap([&](const PGMap& pg_map) {
+	if (pg_map.num_pg_unknown > 0) {
+	  ss << pg_map.num_pg_unknown << " pgs have unknown state; cannot draw"
+	     << " any conclusions";
+	  r = -EAGAIN;
+	  return;
+	}
+	auto q = pg_map.num_pg_by_osd.find(osd);
+	if (q != pg_map.num_pg_by_osd.end()) {
+	  if (q->second.acting > 0 || q->second.up > 0) {
+	    ss << "currently " << q->second.acting + q->second.up
+	       << " pgs map to osd." << osd << " as acting or up";
+	    r = -EBUSY;
+	    return;
+	  }
+	}
+	int num_active_clean = 0;
+	for (auto& p : num_pg_by_state) {
+	  if (p.first.find("active+clean") == 0) {
+	    num_active_clean += p.second;
+	  }
+	}
+	if (num_active_clean < pg_map.num_pg) {
+	  // all pgs aren't active+clean; we need to be careful.
+	  auto p = pg_map.osd_stat.find(osd);
+	  if (p == pg_map.osd_stat.end()) {
+	    ss << "osd." << osd << " has no reported stats; cannot draw any"
+	       << " conclusions";
+	    r = -EAGAIN;
+	    return;
+	  }
+	  if (p->second.num_pgs > 0) {
+	    ss << "osd." << osd << " last reported it had " << p->second.num_pgs
+	       << " pgs stored";
+	    r = -EBUSY;
+	    return;
+	  }
+	}
+      });
+    if (r) {
+      cmdctx->reply(r, ss);
+      return true;
+    }
+    ss << "osd." << osd << " is safe to remove";
```

Comment:
Maybe s/remove/destroy/ to keep that verb consistent? (Also consistent with the ceph-disk command)